proposed/2021/ShowVulnerabilitiesInThePMUI.md

```diff
@@ -0,0 +1,90 @@
+
+* Status: **MVP Implementation**
+* Author(s): [Chris Gill](https://github.com/chgill-MSFT)
+* Issue: 
+* Type: Feature
+
+## Description: What is it?
+
+When a vulnerability in a NuGet package is discovered, surface an indicator of the vulnerability in the Package Manager UI to alert developers and help them take the appropriate action.
+
+## Problem: What problem is this solving?
+
+Vulnerabilities in packages can be leveraged by malicious actors to do harm to developers and their users. At the time of writing this, there are 113 NuGet advisories in the [GitHub Advisory Database](https://github.com/advisories?query=ecosystem%3Anuget). While package vulnerabilities can be found on NuGet.org and with the dotnet list command, most developer interactions with NuGet occur in the Visual Studio package manager UI where no vulnerability information is currently available.
+
+## Why: How do we know this is a real problem worth solving?
+
+NuGet hosts over 200,000 packages that developers use today. Most packages depend on another package, to which the attack vector / blast radius of any security vulnerability can be catastrophic in any package management ecosystem.
+
+## Success: How do we know if we've solved this problem?
+
+We will use the following to measure the success of this feature:
+
+* We see a significant decrease in direct installations for vulnerable packages versions after surfacing vulnerability indicators in the PMUI.
+* We see a significant increase in updates and/or uninstalls for directly installed vulnerable package versions after surfacing vulnerability indicators in the PMUI.
+    * After we enable customers to view transitively installed packages, we will also measure this update/uninstall rates for vulnerable transitive packages.
+* Qualitative feedback from customer interviews following the release of this feature show that customers find this feature very helpful for securing their NuGet packages.
+
+## Audience: Who are we building for?
+
+We are building this for Visual Studio users who manage and install packages through the package manager UI.
+
+## Scope
+
+**The MVP of this feature will be focused on enabling the following scenarios for Visual Studio .NET developers:**
+
+- Users can discover when they have top level packages that have a known vulnerabilities in the PMUI.
+- Users can see when a package search result in the Browse tab has a known vulnerability, to help them make informed decisions when choosing packages.
+- Users can find which installed top level packages have a vulnerability, to help them take the appropriate action.
+- Minor changes to the deprecation UX to align with necessary vulnerability UX.
+
+The following scenarios will not be in the MVP, but will be considered for next iterations:
+
+- Users can discover if they have installed packages with vulnerabilities through indicators in the solution explorer.
+- Users can see which transitively installed packages have vulnerabilities.
```

Comment:
This (potential) omission raised some concerns with @anangaur when we took an earlier pass at this work, and it would be good to confirm how this is viewed now. The main concern is that to include vulnerabilities only on top level packages has the potential to give false negatives to users when in fact they do have vulnerable transitive packages installed. This was a trust concern--we didn't want customers to think they didn't have vulnerabilities when they do, and to present some form of this feature only showing top level vulnerabilities had the potential to do that.

On the other side of this, we have the need to sometimes populate a large graph if the transitive tree is sizeable, and this could raise perf issues. One way to approach this could be a background load of metadata, with clear visual cues to the customer that the full vulnerabilities/deprecations picture is still pending (e.g. a progress wheel in the tab or the like).

Reply:
We decided to prioritize viewing top level vulnerabilities ahead of transitive dependencies. The rationale being that catching some vulnerabilities is better than catching none. We'll make it clear in the announcing blogpost that this feature is limited to top level dependencies for the time being so we can get feedback early on and iterate.